src/main/java/com/github/pgasync/PgConnectionPool.java

```diff
@@ -0,0 +1,379 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.github.pgasync;
+
+import com.pgasync.Connection;
+import com.pgasync.Listening;
+import com.pgasync.PreparedStatement;
+import com.pgasync.Row;
+import com.pgasync.SqlException;
+import com.pgasync.NettyConnectibleBuilder;
+import com.pgasync.ResultSet;
+import com.pgasync.Transaction;
+
+import javax.annotation.concurrent.GuardedBy;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Resource pool for backend connections.
+ *
+ * @author Antti Laisi
+ */
+public class PgConnectionPool extends PgConnectible {
+
+    private class PooledPgConnection implements Connection {
+
+        private class PooledPgTransaction implements Transaction {
+
+            private final Transaction delegate;
+
+            PooledPgTransaction(Transaction delegate) {
+                this.delegate = delegate;
+            }
+
+            public CompletableFuture<Void> commit() {
+                return delegate.commit();
+            }
+
+            public CompletableFuture<Void> rollback() {
+                return delegate.rollback();
+            }
+
+            public CompletableFuture<Void> close() {
+                return delegate.close();
+            }
+
+            public CompletableFuture<Transaction> begin() {
+                return delegate.begin().thenApply(PooledPgTransaction::new);
+            }
+
+            @Override
+            public CompletableFuture<Integer> query(BiConsumer<Map<String, PgColumn>, PgColumn[]> onColumns, Consumer<Row> onRow, String sql, Object... params) {
+                return delegate.query(onColumns, onRow, sql, params);
+            }
+
+            @Override
+            public CompletableFuture<Void> script(BiConsumer<Map<String, PgColumn>, PgColumn[]> onColumns, Consumer<Row> onRow, Consumer<Integer> onAffected, String sql) {
+                return delegate.script(onColumns, onRow, onAffected, sql);
+            }
+
+            public Connection getConnection() {
+                return PooledPgConnection.this;
+            }
+        }
+
+        private final PgConnection delegate;
+        private PooledPgPreparedStatement evicted;
+        private final LinkedHashMap<String, PooledPgPreparedStatement> statements = new LinkedHashMap<>() {
```

Comment:
LinkedHashMap is not thread safe but is mutated without any locking

Reply:
Check the JavaDoc on `com.pgasync.Connection` interface please. Connections are desined to be never thread safe. So, it is not an issue since `statements` map is a `PgConnection` instance field.

Reply:
Ah, you're correct, I tripped in the deep nesting of classes.